src/Microsoft.Identity.Web/DownstreamApiSupport/DownstreamWebApi.cs

```diff
@@ -0,0 +1,172 @@
+                response = await _httpClient.SendAsync(httpRequestMessage).ConfigureAwait(false);
+            }
+
+            return response;
+        }
+
+        /// <summary>
+        /// Merge the options from configuration and override from caller.
+        /// </summary>
+        /// <param name="optionsInstanceName">Named configuration.</param>
+        /// <param name="calledApiOptionsOverride">Delegate to override the configuration.</param>
+        internal /* for tests */ CalledApiOptions MergeOptions(
+            string optionsInstanceName,
+            Action<CalledApiOptions>? calledApiOptionsOverride)
+        {
+            // Gets the options from configuration (or default value)
+            CalledApiOptions options;
+            if (optionsInstanceName != null)
+            {
+                options = _namedOptions.Get(optionsInstanceName);
+            }
+            else
+            {
+                options = _namedOptions.CurrentValue;
+            }
+
+            // Give a chance to the called to override defaults for this call
+            CalledApiOptions clonedOptions = options.Clone();
+            calledApiOptionsOverride?.Invoke(clonedOptions);
+            return clonedOptions;
+        }
+
+        /// <inheritdoc/>
+        public async Task<TOutput> CallWebApiForUserAsync<TInput, TOutput>(
+            string optionsInstanceName,
+            TInput input,
+            Action<CalledApiOptions>? calledApiOptionsOverride,
+            ClaimsPrincipal? user)
+            where TOutput : class
+        {
+            StringContent? jsoncontent;
+            if (input != null)
+            {
+                var jsonRequest = JsonSerializer.Serialize(input);
+                jsoncontent = new StringContent(jsonRequest, Encoding.UTF8, "application/json");
+// case of patch?               jsoncontent = new StringContent(jsonRequest, Encoding.UTF8, "application/json-patch+json");
+            }
+            else
+            {
+                jsoncontent = null;
+            }
+
+            HttpResponseMessage response = await CallWebApiForUserAsync(
+                optionsInstanceName,
+                calledApiOptionsOverride,
+                user,
+                jsoncontent).ConfigureAwait(false);
+
+            if (response.StatusCode == HttpStatusCode.OK)
+            {
+                string content = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                TOutput? output = JsonSerializer.Deserialize<TOutput>(content, _jsonOptions);
+                return output;
+            }
+            else
+            {
+                string error = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                throw new HttpRequestException($"Invalid status code in the HttpResponseMessage: {response.StatusCode}: {error}");
+            }
+        }
+
+        /// <inheritdoc/>
+        public async Task<HttpResponseMessage> CallWebApiForAppAsync(
```

Comment:
Do we also want to have a generic typed version for this method?

Reply:
It's more work.